strawberryfields/backends/states.py

```diff
@@ -689,6 +744,124 @@ def quad_expectation(self, mode, phi=0, **kwargs):
+            raise ValueError("Matrix of quadratic coefficients A must be of size 2Nx2N.")
+
+        if np.any(A.T != A):
+            raise ValueError("Matrix of quadratic coefficients A must be symmetric.")
+
+        if d is None:
+            linear_coeff = np.zeros([2*self._modes])
+        else:
+            linear_coeff = d.copy()
+            linear_coeff[self._modes:] = -d[self._modes:]
+
+        if linear_coeff.shape != (2*self._modes,):
+            raise ValueError("Vector of linear coefficients d must be of length 2N.")
+
+        # expand the cutoff dimension in approximating the x and p
+        # operators in the Fock basis, to reduce numerical inaccuracy.
+        worksize = kwargs.get('worksize', 5)
+        dim = self._cutoff + worksize
+
+        # construct the x and p operators
+        a = np.diag(np.sqrt(np.arange(1, dim)), 1)
+        x = np.sqrt(self._hbar/2) * (a + a.T)
+        p = -1j * np.sqrt(self._hbar/2) * (a - a.T)
+
+        if phi != 0:
+            # rotate the quadrature operators
+            xold = x.copy()
+            pold = p.copy()
+            x = np.cos(phi)*xold - np.sin(phi)*pold
+            p = np.sin(phi)*xold + np.cos(phi)*pold
+
+        def expand_dims(op, n, modes):
+            """Expand quadrature operator to act on nth mode"""
+            I = np.identity(dim)
+            allowed_indices = zip(indices[:2*modes:2], indices[1:2*modes:2])
+            ind = ','.join(a+b for a, b in allowed_indices)
+            ops = [I]*n + [op] + [I]*(modes-n-1)
+            # the einsum 'ij,kl,mn->ijklmn' (for 3 modes)
+            return np.einsum(ind, *ops)
+
+        # determine modes with quadratic expectation values
+        nonzero = np.concatenate([np.mod(A.nonzero()[0], self._modes), np.mod(linear_coeff.nonzero()[0], self._modes)])
+        ex_modes = list(set(nonzero))
+        num_modes = len(ex_modes)
+
+        if not ex_modes:
+            # only a constant term was provided
+            return k, 0.
+
+        # There are non-zero elements of A and/or d
+        # therefore there are quadratic and/or linear terms.
+        # find the reduced density matrix
+        rho = self.reduced_dm(modes=ex_modes)
+
+        # generate vector of quadrature operators
+        # this array will have shape [2*num_modes] + [dim]*(2*num_modes)
+        r = np.empty([2*num_modes] + [dim]*(2*num_modes), dtype=np.complex128)
+        for n in range(num_modes):
+            r[n] = expand_dims(x, n, num_modes)
+            r[num_modes+n] = expand_dims(p, n, num_modes)
+
+        # reduce the size of A so that we only consider modes
+        # which we need to calculate the expectation value for
+        rows = ex_modes + [i+self._modes for i in ex_modes]
+        quad_coeffs = A[:, rows][rows]/2
+        quad_coeffs[num_modes:, :num_modes] = -quad_coeffs[num_modes:, :num_modes]
+        quad_coeffs[:num_modes, num_modes:] = -quad_coeffs[:num_modes, num_modes:]
```

Comment:
See question above. What is the role of these lines? Why do we make these particular elements of the matrix negative?